pyocd/target/family/target_lpc5500.py

```diff
@@ -132,13 +163,44 @@ def trace_start(self):
         # To prevent this, we explicitly (re)enable traceclk.
         self._enable_traceclk()
 
+    def unlock(self):
+        """! @brief Unlock Cores. See UM11126 51.6.1 """
+        assert self.dp.probe.is_open
+
+        LOG.info("attempting unlock procedure")
+        self.dp.probe.write_dp(addr=0x8, data=0x020000f0)
+        self.dp.probe.flush()
+        LOG.info("read")
+        target_id = self.dp.probe.read_ap(addr=0xc)
+        if target_id != 0x002a0000:
+            LOG.error("could not find AP 2")
+            return
+
+        addr = 0x02000000
+
+        self.dp.probe.write_dp(addr=0x8, data=addr)
+        self.dp.probe.write_ap(addr=addr + 0,   data=0x21)
+        self.dp.probe.flush()
+        sleep(0.1)
+        retval = self.dp.probe.read_ap(addr=addr + 0)
+        if retval != 0:
+            LOG.error("did not reset")
+            return
+        self.dp.probe.write_ap(addr=addr + 4, data=7)
+        self.dp.probe.flush()
+        retval = self.dp.probe.read_ap(addr=addr + 8) & 0xffff
```

Comment:
The AP/DP register accesses should be done using the `DebugPort` methods rather than going through the probe.  You can directly use the AP register access methods on `DebugPort` with a "full" AP register address, such as 0x0200_00FC to read the IDR. The appropriate DP registers will be updated if necessary to select the AP (at offset 24 in the addr) and bank (offset 4).

It's also not necessarily safe to directly call the probe methods, since that bypasses the register cache in `DebugPort`, and it can thus get out of sync.

Reply:
Thanks, I've passed in the AP created by the caller, but there's no flush() operation so that's still called directly.